### PR TITLE
feat(api-abuse): Add rate limit page in Snuba admin

### PR DESCRIPTION
**Pull request by enochtangg**
### Overview
The management of rate limit configs is difficult in Snuba admin. More specifically, viewing applied rate limit configs can get mentally exhausting as rate limit configs are coupled with all runtime configs which might require some line-by-line parsing by on-call. This PR is a first iteration of what a Rate Limit page in Snuba admin might look like. This new page copies the Runtime Config page, removes all non-rate limit configs, and adds some filtering capabilities (by dimension, parameter, target).

rate_limit_faster

### CODE REVIEW

File: snuba/admin/static/ratelimit/index.tsx [hunk 1]
```diff
@@ -0,0 +1,426 @@
...
+  });
+
+  // All supported rate limiter dimensions and parameters in Snuba
+  const availableDimensions = [
+    "project_referrer",
+    "project",
+    "referrer",
+    "organization",
+  ];
+  const availableParameters = ["per_second_limit", "concurrent_limit"];
```
Comment (rahul-kumar-saini) [inline snuba/admin/static/ratelimit/index.tsx:89]: What happens if a key contains some of these keywords but isn't a rate limit config? I can see `project`, `referrer` etc come up in unrelated configs.

File: snuba/admin/static/ratelimit/index.tsx [hunk 2]
```diff
@@ -0,0 +1,426 @@
...
+      }
+    });
+    setFilterTerms(filterTerm);
+
+    let searchData = data
+      .filter((obj) => {
+        if (filterTerm.dimension === "project") {
+          return (
+            obj.key.includes(filterTerm.dimension) &&
+            !obj.key.includes("project_referrer")
```
Comment (rahul-kumar-saini) [inline snuba/admin/static/ratelimit/index.tsx:177]: `"project_referrer"`, `"organization"` etc should probably be constants

**Review by rahul-kumar-saini — COMMENTED** [review]
I feel like most of the rate limit component is the same as the runtime config component so you could refactor out a lot of this code and reuse it in runtime config + ratelimit config

File: snuba/admin/static/ratelimit/index.tsx [hunk 3]
```diff
@@ -0,0 +1,426 @@
...
+    api.getConfigs().then((res) => {
+      let data = filterOnlyRateLimitConfigs(res);
+      populateFilterLists(data);
+      setSearchFilteredData(data);
+      setData(data);
+    });
+  }
+
+  function filterOnlyRateLimitConfigs(configs: Array<Config>) {
+    // Rate limit run time configs must include per_second_limit or concurrent_limit as substring
```
Comment (evanh) [inline snuba/admin/static/ratelimit/index.tsx:110]: You should also include `thread_quota` here.

File: snuba/admin/static/ratelimit/index.tsx [hunk 4]
```diff
@@ -0,0 +1,426 @@
...
+  function fetchData() {
+    api.getConfigs().then((res) => {
+      let data = filterOnlyRateLimitConfigs(res);
+      populateFilterLists(data);
+      setSearchFilteredData(data);
+      setData(data);
+    });
+  }
+
+  function filterOnlyRateLimitConfigs(configs: Array<Config>) {
```
Comment (evanh) [inline snuba/admin/static/ratelimit/index.tsx:109]: I thought we had talked about moving this functionality into the Python backend and either pass a parameter to the config endpoint to only return rate limiting or have a separate endpoint altogether?

**Review by lynnagara — CHANGES_REQUESTED** [review]
Not a fan of this. It duplicates way too much code and even if it did not - makes it seem to the user like rate limit is something special other than config when it is not. Why not just add a simple filter to the page if you don't want to see the rest in this view?

**Comment by evanh** [comment]
> makes it seem to the user like rate limit is something special other than config when it is not.

I don't think this is the right way to think about it. Rate limiting is a distinct use case of configuration, just like feature rollout is another. Separating those two things visually makes perfect sense: why would I want to sort through a bunch of rollout keys in order to check what rate limiting keys I have set, and vice versa?

Configuration in redis is a tool used by many different featrures of Snuba. Forcing every key of every feature to all be in the same place just because they happen to use the same tool is not a good way to manage our system.

**Comment by evanh** [comment]
@lynnagara Would it make more sense if we namespaced the rate limit keys in Redis so the use case is clearly separated?

File: snuba/admin/static/ratelimit/index.tsx (hunk 1 above)
Comment (enochtangg) [reply snuba/admin/static/ratelimit/index.tsx:89]: yeah this is a good point, one thing change we've discussed offline is writing new endpoints which filter rate limit configs in the backend and serve that as a new API route to admin?

File: snuba/admin/static/ratelimit/index.tsx (hunk 4 above)
Comment (enochtangg) [reply snuba/admin/static/ratelimit/index.tsx:109]: We have, this was just the original PR. Probably should have been a draft. Those backend changes are still being made.

**Comment by lynnagara** [comment]
@evanh The part that I'm opposed to is pretending that things are two different systems when they are not. If you think that the rate limit system should be a different one then let's make it a different one rather than put a bunch of filtering on strings (which is incredibly brittle) in front of runtime config.

**Comment by evanh** [comment]
I agree with you about the front end filtering, I wanted to move the logic to the backend. What I'm trying to understand is, what would we have to change to make this "two different systems"? Is your concern just about doing filtering on the front end? What if the filtering was on the backend endpoint?

**Comment by lynnagara** [comment]
If it was filtering alone on the backend endpoint, I think that would not be enough. A lot of the same issues would still remain since they are still the same redis database and namespace. For example: you enter a runtime config and if happens to have a specific format it ends up on the other rate limit page instead. This should never be possible in a separate system.

**Comment by evanh** [comment]
OK, well that will require some deep thought on how to handle the data migration.

@lynnagara Do you think some version of this is worth launching now to help with managing the system? Or is it not worth the tech debt?

**Pull request closed without merging**